src/model.ts

```diff
@@ -17,13 +17,18 @@ export async function getCommitCoverageRatio(
     { repo, rev }: Pick<ResolvedRootURI, 'repo' | 'rev'>,
     endpoint: Endpoint,
     sourcegraph: typeof import('sourcegraph')
-): Promise<number | undefined> {
+): Promise<number | null | undefined> {
     const data = await codecovGetCommitCoverage({
         ...codecovParamsForRepositoryCommit({ repo, rev }, sourcegraph),
         baseURL: endpoint.url,
         token: endpoint.token,
     })
-    return data.commit.totals.coverage
+
+    if (data.commit && data.commit.totals && data.commit.totals.coverage) {
```

Comment:
So the function `codecovGetCommitCoverage`'s return type says that all these properties are non-optional, so by the type declaration, this branch would never be hit. So the types are wrong, I think the case that is happening is CodeCov returns a 404 and then the object is not actually of this type but some kind of JSON error object. I think the right fix here would be to check the `response.ok` property inside `codecovGetCommitCoverage` function to make sure the HTTP status code is 200, and throw an exception, and maybe in addition check `response.statusCode` for `404` specifically and return `null` in that case which would be enforced to be handled by the compiler.

Reply:
Yeah but you're still passing `data.commit.totals.coverage` on a higher level.. Which doesn't exist alot of times.. How would you suggest refactoring this? Because I feel the structure of this wrapping function could be a little cleaner, I'm trying to refactor it right now, but have no clue whether to `return` or `throw` something, and what to return and where to log the message. Can you maybe come in and refactor it quickly because I feel you have more knowledge about this part..

Reply:
@3askaal I've just pushed https://github.com/codecov/sourcegraph-codecov/commit/90ac9e418c4357eaad12241b6607046910839f26 implementing what @felixfbecker describes above.

I guess one thing I'm unclear on is: can the Codecov API return a successful (200) response with a `null` or `undefined` `data.commit.totals.coverage`? In that case we would need to change the  `CodecovCommitData` type to reflect this, and reinstate checks like the ones you had originally put in place.

Reply:
@lguychard, thanks for the help man! And I'm pretty sure this field always exists. So think it's fine..